<h1 align="center">Module 4 Assessment</h1>

## Overview

This assessment is designed to test your understanding of the Mod 4 material. It covers:

* Calculus, Cost Function, and Gradient Descent
* Extensions to Linear Models
* Time Series

Read the instructions carefully. You will be asked both to write code and respond to a few short answer questions.

### Note on the short answer questions

For the short answer questions please use your own words. The expectation is that you have not copied and pasted from an external source, even if you consult another source to help craft your response. While the short answer questions are not necessarily being assessed on grammatical correctness or sentence structure, do your best to communicate yourself clearly.

---
## Part I: Calculus, Cost Function, and Gradient Descent [Suggested Time: 15 min]
---

![best fit line](visuals/best_fit_line.png)

The best fit line that goes through the scatterplot up above can be generalized in the following equation: $$y = mx + b$$

Of all the possible lines, we can prove why that particular line was chosen using the plot down below:

![](visuals/cost_curve.png)

where RSS is defined as the residual sum of squares:

$$ 
\begin{align}
RSS &= \sum_{i=1}^n(actual - expected)^2 \\
&= \sum_{i=1}^n(y_i - \hat{y})^2 \\
&= \sum_{i=1}^n(y_i - (mx_i + b))^2
\end{align}
$$ 

### 1. What is a more generalized name for the RSS curve above? How is it related to machine learning models?

In [1]:
# Your answer here

# The RSS is also known as the method of least squares, and it essentially shows
# us the regression coefficent. We're looking for small residuals becuase
# this coincides with a better r^2 value. Since r^2 tells us how close our data
# is to the best fit line, it's key in evaluating how our model performs.

### 2. Would you rather choose a $m$ value of 0.08 or 0.05 from the RSS curve up above?   What is the relation between the position on the cost curve, the error, and the slope of the line?

In [2]:
# Your answer here

# From this visual it looks as if the gradient descent has found a minimum very close
# to 0.05. If this is the case, then the ideal m value is 0.05 as this will provide the 
# quickest path to the lowest point.

# Gradient descent is kind of like a marble dropping down a stair well-
# it will make its jump to the next steepest point and dropping to it 
# until it reaches the bottom of the curve. If it overshoots (which could
# happen with an m value of 0.08 here) it will go backwards to locate the
# step it missed.

![](visuals/gd.png)

### 3. Using the gradient descent visual from above, explain why the distance between each step is getting smaller as more steps occur with gradient descent.

In [4]:
# Your answer here

# Gradient descent uses derivatives to determine its next step, so 
# as the slope of the function becomes less steep, the iterative step
# that comes next will continue to decrease.

### 4. What is the purpose of a learning rate in gradient descent? Explain how a very small and a very large learning rate would affect the gradient descent.

In [3]:
# Your answer here

# Too small of a learning rate would take much longer to locate the minimum y value,
# too large would overshoot the bottom and result in a lot of back
# and forth. We can also use batch gradient descent to reduce the
# computational expense of the process.

---
## Part 2: Extensions to Linear Regression [Suggested Time: 20 min]
---

In this section, you're going to be creating linear models that are more complicated than a simple linear regression. In the cells below, we are importing relevant modules that you might need later on. We also load and prepare the dataset for you.

In [16]:
import pandas as pd
import itertools
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import numpy as np
from sklearn.linear_model import Lasso, Ridge
import pickle
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import PolynomialFeatures
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error

In [6]:
data = pd.read_csv('raw_data/advertising.csv').drop('Unnamed: 0',axis=1)
data.describe()

TV       radio   newspaper       sales
count  200.000000  200.000000  200.000000  200.000000
mean   147.042500   23.264000   30.554000   14.022500
std     85.854236   14.846809   21.778621    5.217457
min      0.700000    0.000000    0.300000    1.600000
25%     74.375000    9.975000   12.750000   10.375000
50%    149.750000   22.900000   25.750000   12.900000
75%    218.825000   36.525000   45.100000   17.400000
max    296.400000   49.600000  114.000000   27.000000

In [7]:
X = data.drop('sales', axis=1)
y = data['sales']

In [8]:
# split the data into training and testing set. Do not change the random state please!
X_train , X_test, y_train, y_test = train_test_split(X, y,random_state=2019)

### 1. We'd like to add a bit of complexity to the model created in the example above, and we will do it by adding some polynomial terms. Write a function to calculate train and test error for different polynomial degrees.

This function should:
* take `degree` as a parameter that will be used to create polynomial features to be used in a linear regression model
* create a PolynomialFeatures object for each degree and fit a linear regression model using the transformed data
* calculate the mean square error for each level of polynomial
* return the `train_error` and `test_error` 


In [43]:
def polynomial_regression(degree):
    polynomial_features = PolynomialFeatures(degree)
    x_poly = polynomial_features.fit_transform(X_train)
    
    lr = LinearRegression()
    lr.fit(x_poly, y_train)
    y_poly_pred = lr.predict(x_poly)
    
    mse = mean_squared_error(y_train, y_poly_pred)
    
    # just having trouble with simply getting my mse for each
    # data set, below
    
    train_error = None
    test_error = None
    
    # despite that, I can get my overall mse:
    
    return mse

#### Try out your new function

In [44]:
polynomial_regression(3)



0.24235967358392063

### My Answers

MSE for degree 3:

Train: 0.24235967358392063
Test: 0.05386326893830295

MSE for degree 4:

Train: 0.18179109513350636
Test: 0.1498518006243622

My train answer for both degrees is very very close, however my test for each degree is WAY off. My function is
overfitting.

#### Check your answers

MSE for degree 3:
- Train: 0.2423596735839209
- Test: 0.15281375973923944

MSE for degree 4:
- Train: 0.18179109317368244
- Test: 1.9522597174462015

### 2. What is the optimal number of degrees for our polynomial features in this model? In general, how does increasing the polynomial degree relate to the Bias/Variance tradeoff?  (Note that this graph shows RMSE and not MSE.)

<img src ="visuals/rsme_poly_2.png" width = "600">

<!---
fig, ax = plt.subplots(figsize=(7, 7))
degree = list(range(1, 10 + 1))
ax.plot(degree, error_train[0:len(degree)], "-", label="Train Error")
ax.plot(degree, error_test[0:len(degree)], "-", label="Test Error")
ax.set_yscale("log")
ax.set_xlabel("Polynomial Feature Degree")
ax.set_ylabel("Root Mean Squared Error")
ax.legend()
ax.set_title("Relationship Between Degree and Error")
fig.tight_layout()
fig.savefig("visuals/rsme_poly.png",
            dpi=150,
            bbox_inches="tight")
--->

In [45]:
# Your answer here

# I would keep the degree at 3 or below here.

# Increasing the polynomial degree also increases our chance of overfitting.
# Higher polynomial degreess also means our bias/variance will be more
# out of balance and there will be a higher inverse between the two.

### 3. In general what methods would you can to reduce overfitting and underfitting? Provide an example for both and explain how each technique works to reduce the problems of underfitting and overfitting.

In [46]:
# Your answer here

# The first and in my opinion most foundational way to avpid both underfitting and 
# overfitting is using the right amount of data. Since too little data can 
# cause underfit and too much data can cause overfit, it's important to use
# well balanced, well cleaned, well scaled data from the beginning.

# Cross validation is another ideal way to avoid bad fits, which we can do with
# something like kfold cross validation. Becuase each subset of data here is both
# trained and tested, we reduce the risk of creating different and bad fits for 
# test vs train data.

### 4. What is the difference between the two types of regularization for linear regression?

In [47]:
# Your answer here

# Both L1 and L2 reguilarization seek to simplify the models as much as possible 
# to just get down to the root relationship that our model exposes.

# Ridge regression penalizes high coefficients by forcing them to be as low
# as possible. It can't get rid of features that don't matter but it will attempt
# to minimize their role in the trained model.

# Lasso, on the other hand, punished high value coeffs and sets them to 0 if they
# do not matter. This will reduce our features, which in turn prevents overfit.

### 5. Why is scaling input variables a necessary step before regularization?

In [48]:
# Your answer here

# Without scaling numerical values before regularization, features that have larger numbers or more
# different numbers than most other features may appear to be more important in our model, when in 
# reality they just need to be weighted.

## Time Series: Part 3 [Suggested Time: 15 min]

<!---
# load data
ads_df = pd.read_csv("raw_data/social_network_ads.csv")

# one hot encode categorical feature
def is_female(x):
    """Returns 1 if Female; else 0"""
    if x == "Female":
        return 1
    else:
        return 0
        
ads_df["Female"] = ads_df["Gender"].apply(is_female)
ads_df.drop(["User ID", "Gender"], axis=1, inplace=True)
ads_df.head()

# separate features and target
X = ads_df.drop("Purchased", axis=1)
y = ads_df["Purchased"]

# train/test split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=19)

# preprocessing
scale = StandardScaler()
scale.fit(X_train)
X_train = scale.transform(X_train)
X_test = scale.transform(X_test)

# save preprocessed train/test split objects
pickle.dump(X_train, open("write_data/social_network_ads/X_train_scaled.pkl", "wb"))
pickle.dump(X_test, open("write_data/social_network_ads/X_test_scaled.pkl", "wb"))
pickle.dump(y_train, open("write_data/social_network_ads/y_train.pkl", "wb"))
pickle.dump(y_test, open("write_data/social_network_ads/y_test.pkl", "wb"))

# build model
model = LogisticRegression(C=1e5, solver="lbfgs")
model.fit(X_train, y_train)
y_test_pred = model.predict(X_test)
y_train_pred = model.predict(X_train)

from sklearn.metrics import confusion_matrix

# create confusion matrix
# tn, fp, fn, tp
cnf_matrix = confusion_matrix(y_test, y_test_pred)
cnf_matrix

# build confusion matrix plot
plt.imshow(cnf_matrix,  cmap=plt.cm.Blues) #Create the basic matrix.

# Add title and Axis Labels
plt.title('Confusion Matrix')
plt.ylabel('True label')
plt.xlabel('Predicted label')

# Add appropriate Axis Scales
class_names = set(y_test) #Get class labels to add to matrix
tick_marks = np.arange(len(class_names))
plt.xticks(tick_marks, class_names)
plt.yticks(tick_marks, class_names)

# Add Labels to Each Cell
thresh = cnf_matrix.max() / 2. #Used for text coloring below
#Here we iterate through the confusion matrix and append labels to our visualization.
for i, j in itertools.product(range(cnf_matrix.shape[0]), range(cnf_matrix.shape[1])):
        plt.text(j, i, cnf_matrix[i, j],
                 horizontalalignment="center",
                 color="white" if cnf_matrix[i, j] > thresh else "black")

# Add a Side Bar Legend Showing Colors
plt.colorbar()

# Add padding
plt.tight_layout()
plt.savefig("visuals/cnf_matrix.png",
            dpi=150,
            bbox_inches="tight")
--->

### 1. Which of the following can’t be a component for a time series plot?
A) Seasonality <br>
B) Trend <br>
C) Cyclical <br>
D) Noise<br>
E)  None of the above

In [49]:
# Your answer here

# E, none of the above


### 2) What does autocovariance measure?

A) Linear dependence between multiple points on the different series observed at different times<br>
B) Quadratic dependence between two points on the same series observed at different times<br>
C) Linear dependence between two points on different series observed at same time<br>
D) Linear dependence between two points on the same series observed at different times<br>


In [50]:
# Your answer here

# C) Linear dependence between two points on different series observed at same time

### 3) Looking at the below ACF plot, would you suggest to apply AR or MA in ARIMA modeling technique?

![](visuals/acf.jpg)

A) AR<br>
B) MA <br>
C) Can’t Say <br>



In [51]:
# Your answer here

# B) MA

### 4) Stationarity is a desirable property for a time series process.

A) TRUE <br>
B) FALSE <br>

In [10]:
# Your answer here

### 5) Which of the following statement is correct?

1. If autoregressive parameter (p) in an ARIMA model is 1, it means that there is no auto-correlation in the series.

2. If moving average component (q) in an ARIMA model is 1, it means that there is auto-correlation in the series with lag 1.

3. If integrated component (d) in an ARIMA model is 0, it means that the series is not stationary.

A) Only 1 <br>
B) Both 1 and 2 <br>
C) Only 2 <br>
D)  All of the statements <br>



In [53]:
# Your answer here

# C) Only 2

### 6) BIC penalizes complex models more strongly than the AIC. 

A) TRUE <br>
B) FALSE <br>


In [54]:
# Your answer here

# A) TRUE

### 7) How many AR and MA terms should be included for the time series by looking at the above ACF and PACF plots?


![](visuals/acf_pacf.png)


A) AR (1) MA(0) <br>
B) AR(0)MA(1) <br>
C) AR(2)MA(1) <br>
D) AR(1)MA(2) <br>
E) Can’t Say <br>

In [55]:
# Your answer here

# D) AR(1)MA(2)